File: algorithms/src/Ex_445_add_two_numbers_ii/two_stacks.py
"""
Reference:
https://leetcode.com/problems/add-two-numbers-ii/discuss/92623/Easy-O(n)-Java-Solution-using-Stack

Time complexity: O(m+n)
Space complexity: O(n)

"""
from typing import List

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        # boundary checking
        if l1 is None:
            return l2
        if l2 is None:
            return l1

        stack_l1: List[int] = []
        stack_l2: List[int] = []
        itor1 = l1
        itor2 = l2
        while itor1 is not None:
            stack_l1.append(itor1.val)
            itor1 = itor1.next
        while itor2 is not None:
            stack_l2.append(itor2.val)
            itor2 = itor2.next

        current = ListNode(-1)
        carry = 0
        while len(stack_l1) > 0 or len(stack_l2) > 0:
            sum = 0
            if len(stack_l1) > 0:
                sum += stack_l1.pop()
            if len(stack_l2) > 0:
                sum += stack_l2.pop()

            current.val = (sum + carry) % 10
            carry = (sum + carry) // 10

            # make a temporary head node
            temp = ListNode(-1)
            temp.next = current
            current = temp
        if carry > 0:
            current.val = carry

        return current if current.val != -1 else current.next


def print_list(head: ListNode):
    curr = head
    while curr is not None:
        print(curr.val, end=' ')
        curr = curr.next
        if curr is not None:
            print('->', end=' ')


if __name__ == '__main__':
    solu = Solution()
    list1_1 = ListNode(7,
                       ListNode(2,
                                ListNode(4,
                                         ListNode(3, None))))
    list1_2 = ListNode(5,
                       ListNode(6,
                                ListNode(4, None)))
    res1 = solu.addTwoNumbers(list1_1, list1_2)
    print_list(res1)

File: algorithms/src/Ex_445_add_two_numbers_ii/two_stacks.cpp

#include <stack>

using std::cout;
using std::endl;
using std::stack;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(const ListNode* l1, const ListNode* l2) {
        stack<int> s1;
        stack<int> s2;

        // push all digits in l1 and l2 to s1 and s2 respectively
        const ListNode *it1 = l1;
        while (it1 != nullptr) {
            s1.push(it1->val);
            it1 = it1->next;
        }
        const ListNode *it2 = l2;
        while (it2 != nullptr) {
            s2.push(it2->val);
            it2 = it2->next;
        }

        int sum = 0;
        ListNode *list = new ListNode(-1);
        ListNode *head = nullptr;
        while (!s1.empty() || !s2.empty()) {
            if (!s1.empty()) {
                sum += s1.top();
                s1.pop();
            }
            if (!s2.empty()) {
                sum += s2.top();
                s2.pop();
            }

            list->val = sum % 10;
            head = new ListNode(sum / 10, list);
            list = head;
            sum /= 10;
        }

        return list;
    }
};


int main() {
    Solution solu;
    ListNode *l1 = new ListNode(7, new ListNode(2, new ListNode(4, new ListNode(3))));
    ListNode *l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
    ListNode *res = solu.addTwoNumbers(l1, l2);
    // display res
//    for (ListNode *it = res; it != nullptr; it = it->next) {
//        cout << it->val << " -> ";
//    }
    // delete l1, l2, res
    for (ListNode *it = l1; it != nullptr; ) {
        ListNode *temp = it;
        it = it->next;
        temp->next = nullptr;
        delete temp;
    }
    for (ListNode *it = l2; it != nullptr; ) {
        ListNode *temp = it;
        it = it->next;
        temp->next = nullptr;
        delete temp;
    }
    for (ListNode *it = res; it != nullptr; ) {
        ListNode *temp = it;
        it = it->next;
        temp->next = nullptr;
        delete temp;
    }

    return 0;
}

